packages/loopring_v3/contracts/test/AmmPool.sol

```diff
@@ -360,111 +400,340 @@ contract AmmPool is IBlockReceiver {
-                    uint amount = join.maxAmountsIn[i];
-                    ammBalances[i] = ammBalances[i].add(amount);
-                    // Unlock the amount locked for this join
-                    locked[item.owner][tokens[i].addr] = locked[item.owner][tokens[i].addr].sub(join.maxAmountsIn[i]);
-                    // Make the amount unavailable for withdrawing
-                    balance[item.owner][tokens[i].addr] = balance[item.owner][tokens[i].addr].sub(amount);
-                }
-            } else if (item.workType == WorkType.EXIT) {
-                PoolExit memory exit = item.exit;
-                for (uint i = 0; i < tokens.length; i++) {
-                    uint amount = exit.minAmountsOut[i];
-                    ammBalances[i] = ammBalances[i].sub(amount);
-                    // Make the amount available for withdrawing
-                    balance[item.owner][tokens[i].addr] = balance[item.owner][tokens[i].addr].add(amount);
+    function processJoin(
+        Context  memory ctx,
+        PoolJoin memory join,
+        bytes    memory signature
+        )
+        internal
+    {
+        bytes32 poolTxHash = hashPoolJoin(ctx.DOMAIN_SEPARATOR, join);
+        if (signature.length == 0) {
+            require(queue[queuePos].txHash == poolTxHash, "NOT_APPROVED");
+            delete queue[queuePos];
+            queuePos++;
+        } else {
+            require(poolTxHash.verifySignature(join.owner, signature), "INVALID_SIGNATURE");
+        }
+
+        uint poolTotal = totalSupply();
+        uint ratio = BASE;
+        if (poolTotal > 0) {
+            ratio = (join.poolAmountOut * BASE) / poolTotal;
+        } else {
+            // Important for accuracy
+            require(join.poolAmountOut == INITIAL_SUPPLY, "INITIAL_SUPPLY_UNEXPECTED");
+        }
+
+        for (uint i = 0; i < ctx.tokens.length; i++) {
+            uint amount = ctx.ammBalances[i] * ratio / BASE;
+            if (poolTotal == 0) {
+                amount = join.maxAmountsIn[i];
+            }
+            require(amount <= join.maxAmountsIn[i], "LIMIT_IN");
+            if (join.fromLayer2) {
+                TransferTransaction.Transfer memory transfer = ctx._block.readTransfer(ctx.txIdx++);
+                require(transfer.from == join.owner, "INVALID_TX_DATA");
+                require(transfer.toAccountID == accountID, "INVALID_TX_DATA");
+                require(transfer.tokenID == ctx.tokens[i].tokenID, "INVALID_TX_DATA");
+                require(isAlmostEqual(transfer.amount, amount), "INVALID_TX_DATA");
+                require(transfer.fee == 0, "INVALID_TX_DATA");
+                if (signature.length != 0) {
+                    // Now approve this transfer
+                    transfer.validUntil = 0xffffffff;
+                    bytes32 txHash = TransferTransaction.hashTx(ctx.exchangeDomainSeparator, transfer);
+                    exchange.approveTransaction(join.owner, txHash);
```

Comment:
This seems to be a bug: the user who want to join the pool must approve the transfer using his own onchain tx.  Now the operator will be able to move user's fund into the AMM pool if the user authorized this contract as an agent.

Reply:
Yeah, but only if he approved the pool transaction with a signature or by calling `join`? This is why we have agents that can do stuff like this, with their own authorization method for doing transactions for users.